lib/kernel/src/application.erl

```diff
@@ -130,31 +130,53 @@ ensure_all_started(Application) ->
       Started :: [atom()],
       Reason :: term().
 ensure_all_started(Application, Type) ->
-    case ensure_all_started(Application, Type, []) of
+    case ensure_all_started([Application], [], Type, []) of
 	{ok, Started} ->
 	    {ok, lists:reverse(Started)};
 	{error, Reason, Started} ->
 	    _ = [stop(App) || App <- Started],
 	    {error, Reason}
     end.
 
-ensure_all_started(Application, Type, Started) ->
-    case start(Application, Type) of
-	ok ->
-	    {ok, [Application | Started]};
-	{error, {already_started, Application}} ->
-	    {ok, Started};
-	{error, {not_started, Dependency}} ->
-	    case ensure_all_started(Dependency, Type, Started) of
+ensure_all_started([App | Apps], OptionalApps, Type, Started) ->
+    case ensure_loaded(App) of
+	{ok, Name} ->
+	    case ensure_started(Name, App, Type, Started) of
 		{ok, NewStarted} ->
-		    ensure_all_started(Application, Type, NewStarted);
+		    ensure_all_started(Apps, OptionalApps, Type, NewStarted);
 		Error ->
 		    Error
 	    end;
+	{error, {"no such file or directory", _} = Reason} ->
+	    case lists:member(App, OptionalApps) of
+		true ->
+		    ensure_all_started(Apps, OptionalApps, Type, Started);
+		false ->
+		    {error, {App, Reason}, Started}
+	    end;
 	{error, Reason} ->
-	    {error, {Application, Reason}, Started}
-    end.
+	    {error, {App, Reason}, Started}
+    end;
+ensure_all_started([], _OptionalApps, _Type, Started) ->
+    {ok, Started}.
 
+ensure_started(Name, App, Type, Started) ->
+    {ok, ChildApps} = get_key(Name, applications),
+    {ok, OptionalApps} = get_key(Name, optional_applications),
+
+    case ensure_all_started(ChildApps, OptionalApps, Type, Started) of
```

Comment:
`ensure_all_started/4` does not protects itself against cyclic dependences which would seem to cause an infinite recursion.
Seem to me that a dependency cycle could "hide" in certain combination of available optional applications.

Reply:
My understanding is that this behaviour is unchanged in this patch because OptionalApplications are only taken into account if the `.app` file cannot be found on disk. In case of cyclic apps, assuming both of them are available, then you ran into a loop as we will start both. If one of them is not available, then it won't be started, but by then you don't have a cycle either (as one part is missing).

Reply:
I agree that behaviour is unchanged and existing system will not be affected.
My concern was that with the introduction of optional applications it would be possible to construct a system with several optional applications with a "hidden" cycle that is not detected until a certain combination of optional applications are available together. Maybe I'm overstating this risk as I can't really come up with a simple enough example :-)